src/inmanta/server/services/compilerservice.py

```diff
@@ -517,7 +517,9 @@ async def _cleanup(self) -> None:
         )
         LOGGER.info("Cleaning up compile reports that are older than %s", oldest_retained_date)
         try:
-            await data.Compile.delete_older_than(oldest_retained_date)
+            # Shield this task from cancellation to make sure locks will be released properly if the encasing task
+            # is cancelled in flight. https://github.com/inmanta/inmanta-core/issues/4384
+            await shield(data.Compile.delete_older_than(oldest_retained_date))
```

Comment:
Not sure if this should be done for some other database query as well.

Reply:
I'm not 100% sure either but it looks fine to me. Let's merge it now and I'll make sure to point Arnaud here when he's back.

Reply:
The problem is a lot bigger than this. Each background task that interacts with the database can conflict with the truncate query when it's cancelled. The above-mentioned query is one such background task, but we have several others as well. I think it would be good to do the shielding at the lowest in the database layer. That way all queries are protected.

An alternative I think about now, is to add support to the Scheduler to await tasks when the scheduler is stopped instead of cancelling them. We already have the equivalent functionality in the `add_background_task()` method using the `cancel_on_stop` argument. The advantage of this solution is that all queries are finished by the time we shutdown the DatabaseService. By shielding the task, it keeps executing. That way we might end up in the scenario where the database connection is shut down before the task ends, which again could cause other issues. But I would like to get the opinion of Wouter before we start implementing this. I remember he had a strong opinion on how this should be implemented. But I don't exactly remember the rationale.

Reply:
I lforgot we have the 'cancel on stop' option. 

So it would seem like a good solution to set 'cancel on stop' to false on all the database queries, this makes the shutdown wait for them.